shared/src/main/scala/fetch.scala

```diff
@@ -213,163 +216,163 @@ object `package` {
     /**
      * Lift a plain value to the Fetch monad.
      */
-    def pure[A](a: A): Fetch[A] =
-      Unfetch(IO.pure(Done(a)))
+    def pure[F[_]: Monad, A](a: A): Fetch[F, A] =
+      Unfetch(Monad[F].pure(Done(a)))
 
-    def exception[A](e: Env => FetchException): Fetch[A] =
-      Unfetch(IO.pure(Throw[A](e)))
+    def exception[F[_]: Monad, A](e: Env[F] => FetchException[F]): Fetch[F, A] =
+      Unfetch(Monad[F].pure(Throw[F, A](e)))
 
-    def error[A](e: Throwable): Fetch[A] =
+    def error[F[_]: Monad, A](e: Throwable): Fetch[F, A] =
       exception((env) => UnhandledException(e, env))
 
-    def apply[I, A](id: I, ds: DataSource[I, A])(
+    def apply[F[_]: Concurrent, I, A](id: I, ds: DataSource[F, I, A])(
       implicit
-        CS: ContextShift[IO]
-    ): Fetch[A] =
-      Unfetch(
+        CS: ContextShift[F]
+    ): Fetch[F, A] =
+      Unfetch[F, A](
         for {
-          deferred <- Deferred[IO, FetchStatus]
+          deferred <- Deferred[F, FetchStatus]
           request = FetchOne(id, ds)
           result = deferred.complete _
           blocked = BlockedRequest(request, result)
-          anyDs = ds.asInstanceOf[DataSource[Any, Any]]
+          anyDs = ds.asInstanceOf[DataSource[F, Any, Any]]
           blockedRequest = RequestMap(Map(anyDs -> blocked))
-        } yield Blocked(blockedRequest, Unfetch(
-          deferred.get.flatMap(_ match {
+        } yield Blocked(blockedRequest, Unfetch[F, A](
+          deferred.get.flatMap {
             case FetchDone(a: A) =>
-              IO.pure(Done(a))
+              Monad[F].pure(Done(a))
             case FetchMissing() =>
-              IO.pure(Throw((env) => MissingIdentity(id, request, env)))
-          })
+              Monad[F].pure(Throw((env) => MissingIdentity(id, request, env)))
+          }
         ))
       )
 
     /**
       * Run a `Fetch`, the result in the `IO` monad.
       */
-    def run[A](
-      fa: Fetch[A],
-      cache: DataSourceCache = InMemoryCache.empty
+    def run[F[_]: Sync: Par, A](
+      fa: Fetch[F, A],
+      cache: DataSourceCache[F] = InMemoryCache.empty[F]
```

Comment:
This doesn't compile because the call to `.empty` requires the implicit `Sync` instance, which is in the implicit argument list that follows, but that is not important to the concept.